plumbing/reference.go

```diff
@@ -204,6 +204,20 @@ func (r *Reference) Strings() [2]string {
 }
 
 func (r *Reference) String() string {
-	s := r.Strings()
-	return fmt.Sprintf("%s %s", s[1], s[0])
+	var v strings.Builder
+
+	switch r.Type() {
+	case HashReference:
+		h := r.Hash().String()
+		v.Grow(len(h) + len(r.Name().String()) + 1)
+		v.WriteString(h)
+	case SymbolicReference:
+		s := symrefPrefix + r.Target().String()
+		v.Grow(len(s) + len(r.Name().String()) + 1)
+		v.WriteString(s)
+	}
+
+	v.WriteString(" ")
```

Comment:
As `r.Type()` can theoretically have a `InvalidReference` value. Wouldn't it be more accurate to move this in-line of the `case` checks (which already do take the `+ 1` into account)?

Reply:
@hiddeco Thank you for the feedback. I revisited these changes taking that into account and I think it resulted in a clearer code. The changes did not result in increased allocations nor B/op.

PTAL